File: Java/src/htmlwalker/TagTraceWalker.java
package htmlwalker;

import htmlwalker.core.EventWalker;

public class TagTraceWalker extends EventWalker
{
    private int traceIndex = 0;
    private int traceIndent = 0;

    private void traceBegin() { traceIndent++; }
    private void traceEnd() { traceIndent--; }

    private void trace(int verbosity, ITag tag)
    {
        if (this.verbosity >= verbosity)
        {
            String msg
                = Integer.toString(this.traceIndex) + ":"
                + new String(new char[this.traceIndent]).replace("\0", " ")
                + "<" + tag.name() + ">[" + (tag.getTextValue() == null ? tag.toString() : "") + "]";
            System.out.println(msg);
        }
        this.traceIndex++;
    }

    @SuppressWarnings("unused")
	private void trace(int verbosity, String str)
    {
        if (this.verbosity >= verbosity && str != null && !str.isEmpty())
        {
            String msg = Integer.toString(this.traceIndex) + ":" + new String(new char[this.traceIndent]).replace("\0", " ") + str;
            System.out.println(msg);
        }
        traceIndex++;
    }

    public int verbosity = 1;

    public TagTraceWalker() { super(); }

    @Override
    protected void onStartDocument(DocumentTag adapter) { }

    @Override
    protected void onEndDocument(DocumentTag adapter) { }

    @Override
    protected void onStartTag(ITag adapter)
    {
        traceBegin();
        trace(1, adapter);
    }

    @Override
    protected void onEndTag(ITag adapter) { traceEnd(); }

    @Override
    protected void onTag(ITag adapter) { trace(1, adapter); }
}

File: CSharp/HtmlWalker/TagTraceWalker.cs
using HtmlWalker.Core;
using System;
using System.Diagnostics;

namespace HtmlWalker
{
    // TODO: Make into a modifier
    public class TagTraceWalker : EventWalker
    {
        public static void StartTagTrace(object sender, EventArgs args)
        {
            if (sender is TagTraceWalker tracer && args is TagEventArgs tagArgs)
            {
                tracer.TraceBegin();
                tracer.Trace(tagArgs.Tag);
            }
        }

        public static void EndTagTrace(object sender, EventArgs args)
        {
            if (sender is TagTraceWalker tracer)
                tracer.TraceEnd();
        }

        public static void TagTrace(object sender, EventArgs args)
        {
            if (sender is TagTraceWalker tracer && args is TagEventArgs tagArgs)
                tracer.Trace(tagArgs.Tag);
        }

        #region Tracing

        private int traceIndex = 0;
        private int traceIndent = 0;

        private void TraceBegin() { traceIndent++; }
        private void TraceEnd() { traceIndent--; }

        private void Trace(ITag tag)
        {
            if (Verbosity >= 1)
            {
                var msg = traceIndex.ToString() + ":" + new string(' ', traceIndent) + "<" + tag.Name + ">";
                if (!String.IsNullOrWhiteSpace(tag.TextValue))
                    msg += "[" + tag.TextValue + "]";
                if (DebugTrace)
                    Debug.Print(msg);
                else
                    System.Console.WriteLine(msg);
            }
            traceIndex++;
        }

        private void Trace(int verbosity, string str)
        {
            if (Verbosity >= verbosity && !String.IsNullOrWhiteSpace(str))
            {
                var msg = traceIndex.ToString() + ":" + new string(' ', traceIndent) + str;
                if (DebugTrace)
                    Debug.Print(msg);
                else
                    Console.WriteLine(msg);
            }
            traceIndex++;
        }

        #endregion

        public int Verbosity = 1;

        public bool DebugTrace = false;

        public TagTraceWalker() : base()
        {
            StartTagHandler += StartTagTrace;
            EndTagHandler += EndTagTrace;
            TagHandler += TagTrace;
        }
    }
}
